sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala

```diff
@@ -2434,829 +2447,622 @@ class CollationSQLExpressionsSuite
+          sqlState = Some("42K09"),
+          parameters = Map(
+            "functionName" -> "`collect_set`",
+            "dataType" -> "\"MAP\" or \"COLLATED STRING\"",
+            "sqlExpr" -> "\"collect_set(col)\""),
+          context = ExpectedContext(
+            fragment = "collect_set(col)",
+            start = 7,
+            stop = 22))
+      }
     }
   }
 
   test("element_at supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT element_at(array('a', 'b', 'c'), 2);"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row("b")
-        )
+    testAdditionalCollations.foreach { collation =>
+      val query = "SELECT element_at(array('a', 'b', 'c'), 2);"
+      val result = Seq(
+        Row("b")
       )
-      // check result row data type
       val dataType = StringType(collation)
-      assert(sql(query).schema.head.dataType == dataType)
+      testCollationSqlExpressionCommon(query, collation, result, dataType)
     }
   }
 
   test("aggregate supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT aggregate(array('a', 'b', 'c'), '', (acc, x) -> concat(acc, x));"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row("abc")
-        )
+    testAdditionalCollations.foreach { collation =>
+      val query = "SELECT aggregate(array('a', 'b', 'c'), '', (acc, x) -> concat(acc, x));"
+      val result = Seq(
+        Row("abc")
       )
-      // check result row data type
       val dataType = StringType(collation)
-      assert(sql(query).schema.head.dataType == dataType)
+      testCollationSqlExpressionCommon(query, collation, result, dataType)
     }
   }
 
   test("explode supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT explode(array('a', 'b'));"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row("a"),
-          Row("b")
-        )
+    testAdditionalCollations.foreach { collation =>
+      val query = "SELECT explode(array('a', 'b'));"
+      val result = Seq(
+        Row("a"),
+        Row("b")
       )
-      // check result row data type
       val dataType = StringType(collation)
-      assert(sql(query).schema.head.dataType == dataType)
+      testCollationSqlExpressionCommon(query, collation, result, dataType)
     }
   }
 
   test("posexplode supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT posexplode(array('a', 'b'));"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row(0, "a"),
-          Row(1, "b")
+    testAdditionalCollations.foreach { collation =>
+      val query = "SELECT posexplode(array('a', 'b'));"
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
```

Comment:
Can't you re-use your new function here?

Reply:
@MaxGekk, yes, I've refactored the new function and all the tests to pass the sequence of expected types as an argument, which covered the remaining test cases, so hopefully it should be much more readable now.